ambry-router/src/main/java/com.github.ambry.router/GCMCryptoService.java

```diff
@@ -78,6 +98,52 @@ public ByteBuffer encrypt(ByteBuffer toEncrypt, SecretKeySpec key) throws Genera
     }
   }
 
+  @Override
+  public ByteBuf encrypt(ByteBuf toEncrypt, SecretKeySpec key) throws GeneralSecurityException {
+    return encrypt(toEncrypt, key, false);
+  }
+
+  /**
+   * Helper function to encrypt ByteBuf with the given key. When useFixedIv is true, don't use a random iv byte
+   * array, using a all zero byte array instead. Only set it to be true in test.
+   * @param toEncrypt {@link ByteBuf} that needs to be encrypted
+   * @param key the secret key (of type T) to use to encrypt
+   * @return the {@link ByteBuf} containing the encrypted content. Ensure the result has all
+   * the information like the IV along with the encrypted content, in order to decrypt the content with a given key
+   * @throws {@link GeneralSecurityException} on any exception with encryption
+   */
+  public ByteBuf encrypt(ByteBuf toEncrypt, SecretKeySpec key, boolean useFixedIv) throws GeneralSecurityException {
+    try {
+      Cipher encrypter = Cipher.getInstance(GCM_CRYPTO_INSTANCE, "BC");
+      byte[] iv = useFixedIv ? fixedIvVal : new byte[ivValSize];
+      if (!useFixedIv) {
+        random.nextBytes(iv);
+      }
+      encrypter.init(Cipher.ENCRYPT_MODE, key, new IvParameterSpec(iv));
+      int outputSize = encrypter.getOutputSize(toEncrypt.readableBytes());
+      // We need to expose the array address to the encrypter, so here we need to allocate a ByteBuf at heap.
+      ByteBuf encryptedContent =
+          ByteBufAllocator.DEFAULT.heapBuffer(IVRecord_Format_V1.getIVRecordSize(iv) + outputSize);
+      IVRecord_Format_V1.serializeIVRecord(encryptedContent, iv);
+      if (toEncrypt.hasArray()) {
+        int n = encrypter.doFinal(toEncrypt.array(), toEncrypt.arrayOffset() + toEncrypt.readerIndex(),
+            toEncrypt.readableBytes(), encryptedContent.array(),
+            encryptedContent.arrayOffset() + encryptedContent.writerIndex());
+        toEncrypt.readerIndex(toEncrypt.readerIndex() + toEncrypt.readableBytes());
```

Comment:
I wonder if it would be possible to just use the nioBuffers() method on toEncrypt.
Then you could potentially make multiple calls to update (https://docs.oracle.com/javase/7/docs/api/javax/crypto/Cipher.html#update(java.nio.ByteBuffer,%20java.nio.ByteBuffer)) and end it with a call to doFinal (https://docs.oracle.com/javase/7/docs/api/javax/crypto/Cipher.html#doFinal(java.nio.ByteBuffer,%20java.nio.ByteBuffer))

Making the output buffer direct might be more challenging though so that one might have to stay as a heap buffer

Reply:
the actually implementation of those methods is pretty much what we are doing here. If bytebuffer doesn't have array (not heap buffer or read only), then it creates an temporary byte array and fill it with the data from byte buffer. I don't really see memory footprint would be reduced when if we use nioBuffer.

Reply:
Ah ok, that makes sense. I didn't take a look into the impls, only the docs.

Reply:
One follow up question about this. Given how this method has to either access the array or copy the buffer, would this method impl save resources over just using the default method defined in the interface that calls `ByteBuf.nioBuffer()`?